### PR TITLE
MAE-377: Always set offline auto renew when payment plan is selected

### PR DESCRIPTION
**Pull request by erawat**
## Overview
This PR is mainly to change Membership Sign up behaviour from providing an option (checkbox) for user to set offline auto-renew to always set Membership to offline auto-renew when sign up a Membership with payment plan.  

This PR also fix some code style reported by phpcs. 

## Before
Auto-renew offline checkbox was provided on the New Membership form.
screenshot-compuclient-rse-1133-9ca compubox co uk-2020 09 08-09_28_05

## After
Auto-renew offline checkbox has been removed from the New Membership form.
screenshot-compuclient-mae-377-jvz compubox co uk-2020 09 08-09_28_33

## Technical Details
Before the changed, MembeshipOfflineAutoRenewProcessor was checking if the submitted membership is OfflineAutoRenewMemembeship or if Membership is already auto-renewed by looking at the http request then check if the payment plan was used when sign up for the membership. 

Since the form has been removed, and the offline auto-renewal should only be used for payment plan.  We initially check if the membership is using a payment plan or if membership is not already auto-renewed then we process to set membership, recurring contribution, recurring line item to auto review. 

The logic for setting membership to membership, recurring contribution, recurring line item have not been changed in this PR. Only logic for setting auto-renew for non-payment plan has been removed as the auto-renew does not support non-payment plan.

### CODE REVIEW

**Comment by omarabuhussein** [comment]
some notes : 

1- CRM_MembershipExtras_Hook_PostProcess_MembershipOfflineAutoRenewProcessor class contain the following unused methods : **createRecurringSubscriptionLineItems** and **calculateMinimumFrequencyUnitAndInterval** so they should be removed. 

2- I would change the if statement inside CRM_MembershipExtras_Hook_PostProcess_MembershipOfflineAutoRenewProcessor to call the methods directly like this : 

```
if (!$this->isPaymentPlanWithAtLeastOneInstallment() || $this->isMembershipAlreadyAutoRenewed()) {
``` 

storing them in variables is unnecessary duplication.


3- I think that the commits are a little confusing and I would prefer some restructurings, so for example  : 

A-  These two commits should be combined and they should be the first commit : 

https://github.com/compucorp/uk.co.compucorp.membershipextras/pull/288/commits/48772e3fe8407cc3d82b4ef455436dc928804d31
https://github.com/compucorp/uk.co.compucorp.membershipextras/pull/288/commits/39192d8a8fdfeeed7c94d8fc962c5611fb17e78b


B- and the 2nd commit should be these three  combined together : 

https://github.com/compucorp/uk.co.compucorp.membershipextras/pull/288/commits/5cc9c84f2e0e40c43e7bd9a2cc4134225853960a
https://github.com/compucorp/uk.co.compucorp.membershipextras/pull/288/commits/8025feb7f861c5dfd12f1498a2e2856f668f4bbc
https://github.com/compucorp/uk.co.compucorp.membershipextras/pull/288/commits/24187efb4618bfd8adfc8528a5c82d90a0253e11

along with my suggestions in point 1 and 2 above.

C- and the 3rd commit maybe this : 

https://github.com/compucorp/uk.co.compucorp.membershipextras/pull/288/commits/8db688e861a40322c3a45ae99a5973c6a3301d22

D- then maybe this 
https://github.com/compucorp/uk.co.compucorp.membershipextras/pull/288/commits/ed90dbfa4ca98ef0dc75766693f37fac00a7bad5


E-and then the linter fixes :  https://github.com/compucorp/uk.co.compucorp.membershipextras/pull/288/commits/036c64c1cfea2f94f515c79731fb57e5f560da2d

**Review by omarabuhussein — CHANGES_REQUESTED** [review]
some notes : 

1- CRM_MembershipExtras_Hook_PostProcess_MembershipOfflineAutoRenewProcessor class contain the following unused methods : **createRecurringSubscriptionLineItems** and **calculateMinimumFrequencyUnitAndInterval** so they should be removed. 

2- I would change the if statement inside CRM_MembershipExtras_Hook_PostProcess_MembershipOfflineAutoRenewProcessor to call the methods directly like this : 

```
if (!$this->isPaymentPlanWithAtLeastOneInstallment() || $this->isMembershipAlreadyAutoRenewed()) {
``` 

storing them in variables is unnecessary duplication.


3- I think that the commits are a little confusing and I would prefer some restructurings, so for example  : 

A-  These two commits should be combined and they should be the first commit : 

https://github.com/compucorp/uk.co.compucorp.membershipextras/pull/288/commits/48772e3fe8407cc3d82b4ef455436dc928804d31
https://github.com/compucorp/uk.co.compucorp.membershipextras/pull/288/commits/39192d8a8fdfeeed7c94d8fc962c5611fb17e78b


B- and the 2nd commit should be these three  combined together : 

https://github.com/compucorp/uk.co.compucorp.membershipextras/pull/288/commits/5cc9c84f2e0e40c43e7bd9a2cc4134225853960a
https://github.com/compucorp/uk.co.compucorp.membershipextras/pull/288/commits/8025feb7f861c5dfd12f1498a2e2856f668f4bbc
https://github.com/compucorp/uk.co.compucorp.membershipextras/pull/288/commits/24187efb4618bfd8adfc8528a5c82d90a0253e11

along with my suggestions in point 1 and 2 above.

C- and the 3rd commit maybe this : 

https://github.com/compucorp/uk.co.compucorp.membershipextras/pull/288/commits/8db688e861a40322c3a45ae99a5973c6a3301d22

D- then maybe this 
https://github.com/compucorp/uk.co.compucorp.membershipextras/pull/288/commits/ed90dbfa4ca98ef0dc75766693f37fac00a7bad5


E-and then the linter fixes :  https://github.com/compucorp/uk.co.compucorp.membershipextras/pull/288/commits/036c64c1cfea2f94f515c79731fb57e5f560da2d

**Comment by erawat** [comment]
> some notes :
> 
> 1- CRM_MembershipExtras_Hook_PostProcess_MembershipOfflineAutoRenewProcessor class contain the following unused methods : **createRecurringSubscriptionLineItems** and **calculateMinimumFrequencyUnitAndInterval** so they should be removed.
> 
> 2- I would change the if statement inside CRM_MembershipExtras_Hook_PostProcess_MembershipOfflineAutoRenewProcessor to call the methods directly like this :
> 
> ```
> if (!$this->isPaymentPlanWithAtLeastOneInstallment() || $this->isMembershipAlreadyAutoRenewed()) {
> ```
> 
> storing them in variables is unnecessary duplication.
> 
> 3- I think that the commits are a little confusing and I would prefer some restructurings, so for example :
> 
> A- These two commits should be combined and they should be the first commit :
> 
> [48772e3](https://github.com/compucorp/uk.co.compucorp.membershipextras/commit/48772e3fe8407cc3d82b4ef455436dc928804d31)
> [39192d8](https://github.com/compucorp/uk.co.compucorp.membershipextras/commit/39192d8a8fdfeeed7c94d8fc962c5611fb17e78b)
> 
> B- and the 2nd commit should be these three combined together :
> 
> [5cc9c84](https://github.com/compucorp/uk.co.compucorp.membershipextras/commit/5cc9c84f2e0e40c43e7bd9a2cc4134225853960a)
> [8025feb](https://github.com/compucorp/uk.co.compucorp.membershipextras/commit/8025feb7f861c5dfd12f1498a2e2856f668f4bbc)
> [24187ef](https://github.com/compucorp/uk.co.compucorp.membershipextras/commit/24187efb4618bfd8adfc8528a5c82d90a0253e11)
> 
> along with my suggestions in point 1 and 2 above.
> 
> C- and the 3rd commit maybe this :
> 
> [8db688e](https://github.com/compucorp/uk.co.compucorp.membershipextras/commit/8db688e861a40322c3a45ae99a5973c6a3301d22)
> 
> D- then maybe this
> [ed90dbf](https://github.com/compucorp/uk.co.compucorp.membershipextras/commit/ed90dbfa4ca98ef0dc75766693f37fac00a7bad5)
> 
> E-and then the linter fixes : [036c64c](https://github.com/compucorp/uk.co.compucorp.membershipextras/commit/036c64c1cfea2f94f515c79731fb57e5f560da2d)
> 
> also I am going tomorrow to review this PR : #265 again so you can write tests for this one (we need them as part of this PR)

@omarabuhussein  thanks for spotting point 1.  I refactor to code for point 1 and 2 as well as re-structure the git commits as per you suggested. 

I will wait for PR https://github.com/compucorp/uk.co.compucorp.membershipextras/pull/265 to be reviewed and merge to master, so we could ship the changes with unit tests.

File: CRM/MembershipExtras/Hook/PostProcess/MembershipOfflineAutoRenewProcessor.php [hunk 1]
```diff
@@ -72,8 +50,20 @@ private function isOfflineAutoRenewMembership() {
    *
    * @return mixed
    */
-  private function membershipIsAlreadyAutoRenew() {
-    return CRM_Utils_Request::retrieve('membership_is_already_autorenew', 'Int');
+  private function isMembershipAlreadyAutoRenewed() {
```
Comment (MiyaNoctem) [inline CRM/MembershipExtras/Hook/PostProcess/MembershipOfflineAutoRenewProcessor.php:53]: The name of this function may lend itself for confusions, as it does not check if the membership has renewed automatically, as it implies. What it's really doing is checking if the membership is marked to auto-renew. So maybe we can change the method name to `isMembershipAlreadySetToAutoenew()`.

File: tests/phpunit/CRM/MembershipExtras/Hook/PostProcess/MembershipOfflineAutoRenewProcessorTest.php [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+
+    if ($result['count'] > 0) {
+      return array_shift($result['values']);
+    }
+
+    return [];
+  }
+
+  /**
+   *
```
Comment (MiyaNoctem) [inline tests/phpunit/CRM/MembershipExtras/Hook/PostProcess/MembershipOfflineAutoRenewProcessorTest.php:344]: ?

**Review by MiyaNoctem — APPROVED** [review]

**Pull request merged**